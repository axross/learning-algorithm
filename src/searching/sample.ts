export const sortedCharactors = [
  "a",
  "b",
  "c",
  "d",
  "e",
  "f",
  "g",
  "h",
  "i",
  "j",
  "k",
  "l",
  "m",
  "n",
  "o",
  "p",
  "q",
  "r",
  "s",
  "t",
  "u",
  "v",
  "w",
  "x",
  "y",
  "z"
];

export const unsortedCharactors = [
  "c",
  "n",
  "r",
  "j",
  "m",
  "s",
  "b",
  "u",
  "w",
  "y",
  "h",
  "t",
  "v",
  "f",
  "z",
  "g",
  "x",
  "o",
  "k",
  "q",
  "d",
  "e",
  "p",
  "l",
  "a",
  "i"
];
