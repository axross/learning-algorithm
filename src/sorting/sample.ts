import roundRobin from "../test-utility/round-robin.ts";
import randomInteger from "../test-utility/random-integer.ts";

export const charactors = ["a", "b", "c", "d", "e", "f", "g", "h"];

export const samples = roundRobin(charactors).slice(1);

export function getRandomSample() {
  return samples[randomInteger(0, samples.length - 1)];
}

export const staticSample = ["f", "a", "h", "b", "d", "g", "e", "c"];

export const charactorPairs = ["a0", "a1", "b0", "b1", "c0", "c1", "d0", "d1"];

export const charactorPairSamples = roundRobin(["a0", "b0", "c0", "d0"]).reduce(
  (arr, pattern0) => [
    ...arr,
    ...roundRobin(["a1", "b1", "c1", "d1"]).reduce(
      (arr, pattern1) => [...arr, [...pattern0, ...pattern1]],
      [] as string[][]
    )
  ],
  [] as string[][]
);

export const randomCharactorPairSample =
  charactorPairSamples[randomInteger(0, charactorPairSamples.length - 1)];

export const staticCharactorPairSample =
  randomCharactorPairSample[Math.floor((samples.length / 7) * 5)];
