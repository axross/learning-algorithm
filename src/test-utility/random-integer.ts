const randomInteger = (min: number, max: number) =>
  Math.floor(Math.random() * (max - min)) + min;

export default randomInteger;
