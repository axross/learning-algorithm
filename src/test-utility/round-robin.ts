function roundRobin<Element>(
  elements: Element[],
  list?: Element[][]
): Element[][] {
  if (list === undefined) {
    list = elements.map(element => [element]);
  }

  if (list[0].length === elements.length) {
    return list;
  }

  return list.reduce(
    (arr, item) => {
      const newItems = elements
        .filter(element => !item.includes(element))
        .map(element => [...item, element]);

      return [...arr, ...roundRobin(elements, [...newItems])];
    },
    [] as Element[][]
  );
}

export default roundRobin;
