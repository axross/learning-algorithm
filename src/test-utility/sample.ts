function shuffle<Value>(arr: Value[]): Value[] {
  const shuffled = [...arr];

  for (let i = shuffled.length - 1; i >= 0; i--) {
    const rand = Math.floor(Math.random() * (i + 1));

    [shuffled[i], shuffled[rand]] = [shuffled[rand], shuffled[i]];
  }

  return shuffled;
}

export const sampleCharactors = [
  ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 65)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 97))
];

export const shuffledSampleCharactors = shuffle(sampleCharactors);

export const sampleTwoCharactors = [
  ...Array.from({ length: 5 }, (_, i) => [
    String.fromCharCode(i + 48) + "<",
    String.fromCharCode(i + 48) + ">"
  ]),
  ...Array.from({ length: 13 }, (_, i) => [
    String.fromCharCode(i + 65) + "<",
    String.fromCharCode(i + 65) + ">"
  ]),
  ...Array.from({ length: 13 }, (_, i) => [
    String.fromCharCode(i + 97) + "<",
    String.fromCharCode(i + 97) + ">"
  ])
].reduce((array, value) => [...array, ...value], []);

export const shuffledSampleTwoCharactors = [
  ...shuffle([
    ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48) + "<"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 65) + "<"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 97) + "<")
  ]),
  ...shuffle([
    ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48) + ">"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 65) + ">"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 97) + ">")
  ])
];
