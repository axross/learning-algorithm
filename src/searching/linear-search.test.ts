import linearSearch, { LinearSearchStep } from "./linear-search";
import { sampleCharactors } from "../test-utility/sample";

describe("linearSearch", () => {
  const list = sampleCharactors;

  test(`linearSearch(target: "X") finds the index of the target to be ${sampleCharactors.indexOf(
    "X"
  )}`, () => {
    expect(
      linearSearch({
        list: list,
        target: "X"
      })
    ).toBe(sampleCharactors.indexOf("X"));
  });

  test("linearSearch(target: 25000) doesn't find it", () => {
    expect(
      linearSearch({
        list: list,
        target: "🍣"
      })
    ).toBe(-1);
  });

  test('the step snapshot by linearSearch(target: "X") matches with the previous one', () => {
    const steps: LinearSearchStep<string>[] = [];

    linearSearch({
      list: list,
      target: "X",
      onStep: step => steps.push(step)
    });

    expect(steps).toMatchSnapshot();
  });

  test("the maximum caliculation size is O(n)", () => {
    let maximumStepLength = 0;

    for (let i = 0; i < list.length; ++i) {
      const steps: LinearSearchStep<string>[] = [];

      linearSearch({
        list: list,
        target: list[i],
        onStep: step => steps.push(step)
      });

      if (steps.length > maximumStepLength) {
        maximumStepLength = steps.length;
      }
    }

    expect(maximumStepLength).toBe(list.length);
  });

  test("the avarage caliculation size is O((n + 1) / 2)", () => {
    let totalStepLengths = 0;

    for (let i = 0; i < list.length; ++i) {
      const steps: LinearSearchStep<string>[] = [];

      linearSearch({
        list: list,
        target: list[i],
        onStep: step => steps.push(step)
      });

      totalStepLengths = totalStepLengths + steps.length;
    }

    expect(Math.floor(totalStepLengths / list.length)).toBe(
      Math.floor((list.length + 1) / 2)
    );
  });
});
