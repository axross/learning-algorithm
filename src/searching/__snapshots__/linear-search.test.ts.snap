// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linearSearch the step snapshot by linearSearch(target: "X") matches with the previous one 1`] = `
Array [
  Object {
    "index": 0,
    "value": "0",
  },
  Object {
    "index": 1,
    "value": "1",
  },
  Object {
    "index": 2,
    "value": "2",
  },
  Object {
    "index": 3,
    "value": "3",
  },
  Object {
    "index": 4,
    "value": "4",
  },
  Object {
    "index": 5,
    "value": "A",
  },
  Object {
    "index": 6,
    "value": "B",
  },
  Object {
    "index": 7,
    "value": "C",
  },
  Object {
    "index": 8,
    "value": "D",
  },
  Object {
    "index": 9,
    "value": "E",
  },
  Object {
    "index": 10,
    "value": "F",
  },
  Object {
    "index": 11,
    "value": "G",
  },
  Object {
    "index": 12,
    "value": "H",
  },
  Object {
    "index": 13,
    "value": "I",
  },
  Object {
    "index": 14,
    "value": "J",
  },
  Object {
    "index": 15,
    "value": "K",
  },
  Object {
    "index": 16,
    "value": "L",
  },
  Object {
    "index": 17,
    "value": "M",
  },
  Object {
    "index": 18,
    "value": "N",
  },
  Object {
    "index": 19,
    "value": "O",
  },
  Object {
    "index": 20,
    "value": "P",
  },
  Object {
    "index": 21,
    "value": "Q",
  },
  Object {
    "index": 22,
    "value": "R",
  },
  Object {
    "index": 23,
    "value": "S",
  },
  Object {
    "index": 24,
    "value": "T",
  },
  Object {
    "index": 25,
    "value": "U",
  },
  Object {
    "index": 26,
    "value": "V",
  },
  Object {
    "index": 27,
    "value": "W",
  },
  Object {
    "index": 28,
    "value": "X",
  },
]
`;
